examples/01_icetray/04_i3_module_in_native_icetray_example.py

```diff
@@ -0,0 +1,167 @@
+truth = TRUTH.UPGRADE
+
+
+def construct_mock_model() -> StandardModel:
+    """Construct a mock model for the example.
+
+    Replace this with a trained model.
+    """
+    detector = IceCubeUpgrade(
+        graph_builder=KNNGraphBuilder(nb_nearest_neighbours=8),
+    )
+    gnn = DynEdge(
+        nb_inputs=detector.nb_outputs,
+        global_pooling_schemes=["min", "max", "mean", "sum"],
+    )
+    task = EnergyReconstruction(
+        hidden_size=gnn.nb_outputs,
+        target_labels=["energy"],
+        loss_function=LogCoshLoss(),
+        transform_prediction_and_target=torch.log10,
+    )
+    model = StandardModel(
+        detector=detector,
+        gnn=gnn,
+        tasks=[task],
+        optimizer_class=Adam,
+        optimizer_kwargs={"lr": 1e-03, "eps": 1e-03},
+    )
+    return model
+
+
+def construct_modules(
+    model_dict: Dict[str, Dict], gcd_file: str
+) -> List[I3InferenceModule]:
+    """Construct a list of I3InfereceModules for the I3Deployer."""
+    features = FEATURES.UPGRADE
+    deployment_modules = []
+    for model_name in model_dict.keys():
+        model_path = model_dict[model_name]["model_path"]
+        prediction_columns = model_dict[model_name]["prediction_columns"]
+        pulsemap = model_dict[model_name]["pulsemap"]
+        extractor = I3FeatureExtractorIceCubeUpgrade(pulsemap=pulsemap)
+        deployment_modules.append(
+            I3InferenceModule(
+                pulsemap=pulsemap,
+                features=features,
+                pulsemap_extractor=extractor,
+                model=model_path,
+                gcd_file=gcd_file,
+                prediction_columns=prediction_columns,
+                model_name=model_name,
+            )
+        )
+    return deployment_modules
+
+
+def apply_to_files(
+    i3_files: List[str],
+    gcd_file: str,
+    output_folder: str,
+    modules: Sequence[GraphNeTI3Module],
+) -> None:
+    """Will start an IceTray read/write chain with graphnet modules.
+
+    If n_workers > 1, this function is run in parallel n_worker times. Each
+    worker will loop over an allocated set of i3 files. The new i3 files will
+    appear as copies of the original i3 files but with reconstructions added.
+    Original i3 files are left untouched.
+    """
+    for i3_file in i3_files:
+        tray = I3Tray()
+        tray.context["I3FileStager"] = dataio.get_stagers()
+        tray.AddModule(
+            "I3Reader",
+            "reader",
+            FilenameList=[gcd_file, i3_file],
+        )
+        for i3_module in modules:
+            tray.AddModule(i3_module)
+        tray.Add(
+            "I3Writer",
+            Streams=[
+                icetray.I3Frame.DAQ,
+                icetray.I3Frame.Physics,
+                icetray.I3Frame.TrayInfo,
+                icetray.I3Frame.Simulation,
+            ],
+            filename=output_folder + "/" + i3_file.split("/")[-1],
+        )
+        tray.Execute()
+        tray.Finish()
+    return
```

Comment:
Don't think it's necessary to wrap this in a function.
```suggestion
```

Reply:
I think it gives a nice break between graphnet-specfic things and icetray functionality. So unless you feel this strongly, I'd prefer to keep it this way.

Reply:
I'll leave that to you. :)  I'd at least update the docstring to remove the reference to `n_workers`. ;)